#!/bin/bash

# 
# This is the simplest variant-calling file in here.
# I don't neessarily have the refefence files here ready to run this, but this 
# shows how tidy a script could be.
#


set -euo pipefail

# The name of the reference.
REF=refs/ucsc_GRCh37_chr15_exons.fasta
OUTDIR=results/play1

# The name of the BAM file
BAM=${OUTDIR}/align.bam

# Create a bwa index for the reference.
bwa index $REF

# Create a samtools index for the reference.
samtools faidx $REF

# Simulate reads from the reference file.
dwgsim -d 324 -F 0.2 -R 0.2 -X 0.8 $REF ${OUTDIR}/simulated

# This is the data naming generated by dwgsim.
R1=${OUTDIR}/simulated.bwa.read1.fastq
R2=${OUTDIR}/simulated.bwa.read2.fastq

#
# Generate the alignment from the data simulated above.
#
bwa mem $REF $R1 $R2 | samtools sort > $BAM

# Index the BAM file
samtools index $BAM

# Compute the genotypes from the alignment file.
bcftools mpileup -Ovu -f $REF $BAM > genotypes.vcf

# Call the variants from the genotypes.
bcftools call -vc -Ov genotypes.vcf > observed-mutations.vcf